src/Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration/ReferenceRegistries/SecuritySchemeReferenceRegistry.cs

```diff
@@ -0,0 +1,356 @@
+﻿// ------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All rights reserved.
+//  Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Xml.Linq;
+using Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration.Exceptions;
+using Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration.Extensions;
+using Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration.Models.KnownStrings;
+using Microsoft.OpenApi.Models;
+
+namespace Microsoft.OpenApi.CSharpAnnotations.DocumentGeneration.ReferenceRegistries
+{
+    /// <summary>
+    /// Reference Registry for <see cref="OpenApiSecurityScheme"/>.
+    /// </summary>
+    public class SecuritySchemeReferenceRegistry : ReferenceRegistry<XElement, OpenApiSecurityScheme>
+    {
+        /// <summary>
+        /// The dictionary containing all references of the given security scheme.
+        /// </summary>
+        public override IDictionary<string, OpenApiSecurityScheme> References { get; } =
+            new Dictionary<string, OpenApiSecurityScheme>();
+
+        internal List<string> Scopes { get; private set; }
+
+        /// <summary>
+        /// Finds the existing reference object based on the key from the input or creates a new one.
+        /// </summary>
+        /// <returns>The existing or created reference object.</returns>
+        internal override OpenApiSecurityScheme FindOrAddReference(XElement securityElement)
+        {
+            var type = securityElement.Attribute(KnownXmlStrings.Type)?.Value.Trim();
+            var key = GetKey(securityElement);
+
+            Scopes = new List<string>();
+
+            if (References.ContainsKey(key))
+            {
+                return new OpenApiSecurityScheme
+                {
+                    Reference = new OpenApiReference
+                    {
+                        Id = key,
+                        Type = ReferenceType.SecurityScheme
+                    }
+                };
+            }
+
+            var description = securityElement.Elements().FirstOrDefault(
+                p => p.Name == KnownXmlStrings.Description)?.Value;
+
+            var securityScheme = new OpenApiSecurityScheme
+            {
+                Description = description,
+                Type = GetSecuritySchemeType(type)
+            };
+
+            if (securityScheme.Type == SecuritySchemeType.ApiKey)
```

Comment:
>if [](start = 12, length = 2)

nit, consider switch statement instead. can call private methods that do each scheme type's work to organize the code blocks and make it slightly more modular/readable